// zotero-plugin-toolkit/dist/managers/menu.js
import { ManagerTool } from "../basic.js";
import { UITool } from "../tools/ui.js";
/**
 * Register \<menuitem\>, \<menupopup\>, or \<menuseperator\> to Zotero right-click/window menus.
 */
export class MenuManager extends ManagerTool {
    ui;
    constructor(base) {
        super(base);
        this.ui = new UITool(this);
    }
    /**
     * Insert an menu item/menu(with popup)/menuseprator into a menupopup
     * @remarks
     * options:
     * ```ts
     * export interface MenuitemOptions {
     *   tag: "menuitem" | "menu" | "menuseparator";
     *   id?: string;
     *   label?: string;
     *   // data url (chrome://xxx.png) or base64 url (data:image/png;base64,xxx)
     *   icon?: string;
     *   class?: string;
     *   styles?: { [key: string]: string };
     *   hidden?: boolean;
     *   disabled?: boolean;
     *   oncommand?: string;
     *   commandListener?: EventListenerOrEventListenerObject;
     *   // Attributes below are used when type === "menu"
     *   popupId?: string;
     *   onpopupshowing?: string;
     *   subElementOptions?: Array<MenuitemOptions>;
     * }
     * ```
     * @param menuPopup
     * @param options
     * @param insertPosition
     * @param anchorElement The menuitem will be put before/after `anchorElement`. If not set, put at start/end of the menupopup.
     * @example
     * Insert menuitem with icon into item menupopup
     * ```ts
     * // base64 or chrome:// url
     * const menuIcon = "chrome://addontemplate/content/icons/favicon@0.5x.png";
     * ztoolkit.Menu.register("item", {
     *   tag: "menuitem",
     *   id: "zotero-itemmenu-addontemplate-test",
     *   label: "Addon Template: Menuitem",
     *   oncommand: "alert('Hello World! Default Menuitem.')",
     *   icon: menuIcon,
     * });
     * ```
     * @example
     * Insert menu into file menupopup
     * ```ts
     * ztoolkit.Menu.register(
     *   "menuFile",
     *   {
     *     tag: "menu",
     *     label: "Addon Template: Menupopup",
     *     subElementOptions: [
     *       {
     *         tag: "menuitem",
     *         label: "Addon Template",
     *         oncommand: "alert('Hello World! Sub Menuitem.')",
     *       },
     *     ],
     *   },
     *   "before",
     *   Zotero.getMainWindow().document.querySelector(
     *     "#zotero-itemmenu-addontemplate-test"
     *   )
     * );
     * ```
     */
    register(menuPopup, options, insertPosition = "after", anchorElement) {
        let popup;
        if (typeof menuPopup === "string") {
            // eslint-disable-next-line ts/no-use-before-define
            popup = this.getGlobal("document").querySelector(MenuSelector[menuPopup]);
        }
        else {
            popup = menuPopup;
        }
        if (!popup) {
            return false;
        }
        const doc = popup.ownerDocument;
        const genMenuElement = (menuitemOption) => {
            const elementOption = {
                tag: menuitemOption.tag,
                id: menuitemOption.id,
                namespace: "xul",
                attributes: {
                    label: menuitemOption.label || "",
                    hidden: Boolean(menuitemOption.hidden),
                    disabled: Boolean(menuitemOption.disabled),
                    class: menuitemOption.class || "",
                    oncommand: menuitemOption.oncommand || "",
                },
                classList: menuitemOption.classList,
                styles: menuitemOption.styles || {},
                listeners: [],
                children: [],
            };
            if (menuitemOption.icon) {
                if (!this.getGlobal("Zotero").isMac) {
                    if (menuitemOption.tag === "menu") {
                        elementOption.attributes.class += " menu-iconic";
                    }
                    else {
                        elementOption.attributes.class += " menuitem-iconic";
                    }
                }
                elementOption.styles["list-style-image"] =
                    `url(${menuitemOption.icon})`;
            }
            if (menuitemOption.commandListener) {
                elementOption.listeners?.push({
                    type: "command",
                    listener: menuitemOption.commandListener,
                });
            }
            if (menuitemOption.tag === "menuitem") {
                elementOption.attributes.type = menuitemOption.type || "";
                elementOption.attributes.checked = menuitemOption.checked || false;
            }
            const menuItem = this.ui.createElement(doc, menuitemOption.tag, elementOption);
            if (menuitemOption.isHidden || menuitemOption.getVisibility) {
                popup?.addEventListener("popupshowing", (ev) => {
                    let hidden;
                    if (menuitemOption.isHidden) {
                        hidden = menuitemOption.isHidden(menuItem, ev);
                    }
                    else if (menuitemOption.getVisibility) {
                        const visible = menuitemOption.getVisibility(menuItem, ev);
                        hidden = typeof visible === "undefined" ? undefined : !visible;
                    }
                    if (typeof hidden === "undefined") {
                        return;
                    }
                    if (hidden) {
                        menuItem.setAttribute("hidden", "true");
                    }
                    else {
                        menuItem.removeAttribute("hidden");
                    }
                });
            }
            if (menuitemOption.isDisabled) {
                popup?.addEventListener("popupshowing", (ev) => {
                    const disabled = menuitemOption.isDisabled(menuItem, ev);
                    if (typeof disabled === "undefined") {
                        return;
                    }
                    if (disabled) {
                        menuItem.setAttribute("disabled", "true");
                    }
                    else {
                        menuItem.removeAttribute("disabled");
                    }
                });
            }
            if ((menuitemOption.tag === "menuitem" ||
                menuitemOption.tag === "menuseparator") &&
                menuitemOption.onShowing) {
                popup?.addEventListener("popupshowing", (ev) => {
                    menuitemOption.onShowing(menuItem, ev);
                });
            }
            if (menuitemOption.tag === "menu") {
                const subPopup = this.ui.createElement(doc, "menupopup", {
                    id: menuitemOption.popupId,
                    attributes: { onpopupshowing: menuitemOption.onpopupshowing || "" },
                });
                menuitemOption.children?.forEach((childOption) => {
                    subPopup.append(genMenuElement(childOption));
                });
                menuItem.append(subPopup);
            }
            return menuItem;
        };
        const topMenuItem = genMenuElement(options);
        if (popup.childElementCount) {
            if (!anchorElement) {
                anchorElement = (insertPosition === "after"
                    ? popup.lastElementChild
                    : popup.firstElementChild);
            }
            anchorElement[insertPosition](topMenuItem);
        }
        else {
            popup.appendChild(topMenuItem);
        }
    }
    unregister(menuId) {
        this.getGlobal("document").querySelector(`#${menuId}`)?.remove();
    }
    unregisterAll() {
        this.ui.unregisterAll();
    }
}
var MenuSelector;
(function (MenuSelector) {
    MenuSelector["menuFile"] = "#menu_FilePopup";
    MenuSelector["menuEdit"] = "#menu_EditPopup";
    MenuSelector["menuView"] = "#menu_viewPopup";
    MenuSelector["menuGo"] = "#menu_goPopup";
    MenuSelector["menuTools"] = "#menu_ToolsPopup";
    MenuSelector["menuHelp"] = "#menu_HelpPopup";
    MenuSelector["collection"] = "#zotero-collectionmenu";
    MenuSelector["item"] = "#zotero-itemmenu";
})(MenuSelector || (MenuSelector = {}));
